### PR TITLE
net: lwm2m: Update LwM2M server object to v1.1

### PR DESCRIPTION
**Pull request by jheiskan81**
Update LwM2m server object to v1.1

Added Kconfig configurable choise option for enable Server object version 1.0 or 1.1.
Server Object v1.1 enable by default server initiated bootstrap trigger when CONFIG_LWM2M_RD_CLIENT_SUPPORT_BOOTSTRAP is enabled.

### CODE REVIEW

**Comment by jheiskan81** [comment]
@SeppoTakalo @VeijoPesonen Please review and comments

**Comment by jheiskan81** [comment]
@jarlamsa Please review

File: subsys/net/lib/lwm2m/lwm2m_obj_server.c [hunk 1]
```diff
@@ -21,7 +21,13 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 #endif
 
 #define SERVER_VERSION_MAJOR 1
+#if defined(CONFIG_LWM2M_RD_CLIENT_SUPPORT_BOOTSTRAP)
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_obj_server.c:24]: I don't think this is right, the object version should not depend on whether bootstrap is enabled or not. It should rather be controlled by a separate Kconfig option, like IPSO objects do:
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/lib/lwm2m/Kconfig.ipso#L27

File: subsys/net/lib/lwm2m/lwm2m_obj_server.c [hunk 2]
```diff
@@ -33,8 +39,10 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 #define SERVER_STORE_NOTIFY_ID		6
 #define SERVER_TRANSPORT_BINDING_ID	7
 #define SERVER_REG_UPDATE_TRIGGER_ID	8
+#if defined(CONFIG_LWM2M_RD_CLIENT_SUPPORT_BOOTSTRAP)
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_obj_server.c:42]: This should be controlled by object version.

File: subsys/net/lib/lwm2m/lwm2m_obj_server.c [hunk 3]
```diff
@@ -78,6 +86,10 @@ static struct lwm2m_engine_obj_field fields[] = {
 	/* Mark Transport Binding is RO but BOOTSTRAP needs to write it */
 	OBJ_FIELD_DATA(SERVER_TRANSPORT_BINDING_ID, RW, STRING),
 	OBJ_FIELD_EXECUTE(SERVER_REG_UPDATE_TRIGGER_ID),
+	#if defined(CONFIG_LWM2M_RD_CLIENT_SUPPORT_BOOTSTRAP)
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_obj_server.c:89]: This should be controlled by the object version.

File: subsys/net/lib/lwm2m/lwm2m_rd_client.c [hunk 4]
```diff
@@ -632,6 +651,15 @@ static int sm_do_bootstrap_reg(void)
 {
 	int ret;
 
+	if (client.trigger_bootstrap) {
+		client.trigger_bootstrap = false;
+		client.lifetime = 0U;
+		client.retries = 0U;
+
+		/* All necessary pre work for triggered can be done here before start */
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_rd_client.c:659]: I don't understand this comment, do you expect users to modify this codebase?

File: subsys/net/lib/lwm2m/lwm2m_obj_server.c [hunk 5]
```diff
@@ -33,8 +39,10 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 #define SERVER_STORE_NOTIFY_ID		6
 #define SERVER_TRANSPORT_BINDING_ID	7
 #define SERVER_REG_UPDATE_TRIGGER_ID	8
+#if defined(CONFIG_LWM2M_RD_CLIENT_SUPPORT_BOOTSTRAP)
+#define SERVER_BOOTSTRAP_UPDATE_TRIGGER_ID	9
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_obj_server.c:43]: Also, there seems to be much more resources added in object version 1.1, if we claim to supprot this version we should add them all to the object model.

**Comment by jheiskan81** [comment]
@rlubos Thank you for comment's. I agree use K config for enable Sever Object v1.1 features. I will arrange meeting for discuss more.

**Comment by jheiskan81** [comment]
@rlubos Hi Robert I update a PR by use new Kconfig flag and adding all Server Object v 1.1 resources to PR also. I clean also bootstrap trigger state to enter init state.

**Comment by jheiskan81** [comment]
@SeppoTakalo , @VeijoPesonen , @jarlamsa and @hasnainvirk Please review this.

File: subsys/net/lib/lwm2m/Kconfig [hunk 6]
```diff
@@ -183,6 +183,13 @@ config LWM2M_FIRMWARE_PORT_SECURE
 	help
 	  This is the default server port to connect to for LwM2M firmware downloads over coaps.
 
+config LWM2M_SERVER_OBJECT_VERSION_1_1
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/Kconfig:186]: I'd rather see this as a choice parameter, it sounds more natural for this kind of configs (in case we have more to choose than only 1.0 and 1.1), something like:
```
choice
	prompt "LwM2M Server object version"
	default LWM2M_SERVER_OBJECT_VERSION_1_0
# in the future we could add default LWM2M_SERVER_OBJECT_VERSION_1_1 if LWM2M_VERSION_1_1
	help
	  Select which version of the Server object should be used.

config LWM2M_SERVER_OBJECT_VERSION_1_0
	bool "Server object version 1.0"

config LWM2M_SERVER_OBJECT_VERSION_1_1
	bool "Server object version 1.1"

endchoice
```

File: subsys/net/lib/lwm2m/lwm2m_obj_server.c [hunk 7]
```diff
@@ -33,8 +39,24 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 #define SERVER_STORE_NOTIFY_ID		6
 #define SERVER_TRANSPORT_BINDING_ID	7
 #define SERVER_REG_UPDATE_TRIGGER_ID	8
+#if defined(CONFIG_LWM2M_SERVER_OBJECT_VERSION_1_1)
+#define SERVER_BOOTSTRAP_UPDATE_TRIGGER_ID	9
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_obj_server.c:43]: Would be nice to keep the numbers aligned.

File: subsys/net/lib/lwm2m/lwm2m_obj_server.c [hunk 8]
```diff
@@ -78,6 +100,24 @@ static struct lwm2m_engine_obj_field fields[] = {
 	/* Mark Transport Binding is RO but BOOTSTRAP needs to write it */
 	OBJ_FIELD_DATA(SERVER_TRANSPORT_BINDING_ID, RW, STRING),
 	OBJ_FIELD_EXECUTE(SERVER_REG_UPDATE_TRIGGER_ID),
+	#if defined(CONFIG_LWM2M_SERVER_OBJECT_VERSION_1_1)
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_obj_server.c:103]: The preprocessor directives shouldn't be indented.

**Review by rlubos — COMMENTED** [review]
Looks pretty good, just some nits from my side. Please have a look at the compilance report (git message issues).

File: subsys/net/lib/lwm2m/lwm2m_obj_server.c (hunk 8 above)
Comment (jheiskan81) [reply subsys/net/lib/lwm2m/lwm2m_obj_server.c:103]: OK I will fix that.

File: subsys/net/lib/lwm2m/lwm2m_obj_server.c (hunk 7 above)
Comment (jheiskan81) [reply subsys/net/lib/lwm2m/lwm2m_obj_server.c:43]: OK I will do that.

File: subsys/net/lib/lwm2m/Kconfig (hunk 6 above)
Comment (jheiskan81) [reply subsys/net/lib/lwm2m/Kconfig:186]: Ah Looks good. I will change this.

**Comment by jheiskan81** [comment]
@rlubos OK I fix your comments. Those number alignment looks weird at code review but src code part they are aligned.

File: subsys/net/lib/lwm2m/Kconfig [hunk 9]
```diff
@@ -23,6 +23,21 @@ config LWM2M_DTLS_SUPPORT
 	select NET_SOCKETS_SOCKOPT_TLS
 	select NET_SOCKETS_ENABLE_DTLS
 
+choice
+	prompt "LwM2M Server object version"
+	default LWM2M_SERVER_OBJECT_VERSION_1_0
+# in the future default LWM2M_SERVER_OBJECT_VERSION_1_1 if LWM2M_VERSION_1_1
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/Kconfig:29]: That would just a side comment from my side, I don't think we need this in the code :D

File: subsys/net/lib/lwm2m/lwm2m_obj_server.c [hunk 10]
```diff
@@ -21,20 +21,42 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
...
-#define SERVER_DEFAULT_MIN_PERIOD_ID	2
-#define SERVER_DEFAULT_MAX_PERIOD_ID	3
-#define SERVER_DISABLE_ID		4
-#define SERVER_DISABLE_TIMEOUT_ID	5
-#define SERVER_STORE_NOTIFY_ID		6
-#define SERVER_TRANSPORT_BINDING_ID	7
-#define SERVER_REG_UPDATE_TRIGGER_ID	8
-
-#define SERVER_MAX_ID			9
+#define SERVER_SHORT_SERVER_ID							0
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_obj_server.c:33]: Is your editor configured right? Note Zephyr uses 8 spaces/tab.

Comment (jheiskan81) [reply subsys/net/lib/lwm2m/lwm2m_obj_server.c:33]: Let me check that. Do you want numbers to same line or just add tab?

File: subsys/net/lib/lwm2m/Kconfig (hunk 9 above)
Comment (jheiskan81) [reply subsys/net/lib/lwm2m/Kconfig:29]: OK I will remove this no problem

**Review by rlubos — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

**Comment by VeijoPesonen** [comment]
Would it make sense to mark this as experimental for as long as there are still missing features from v1.1? Please don't take this as a requirement but more like me thinking out loud.

[menuconfig CRYPTO
	bool "Crypto Drivers [EXPERIMENTAL]"
	select EXPERIMENTAL](https://github.com/nrfconnect/sdk-zephyr/blob/c932ad1952970987301719d927fc0232116e908f/drivers/crypto/Kconfig#L9-L11)

**Review by VeijoPesonen — APPROVED** [review]
Looks good to me

File: subsys/net/lib/lwm2m/lwm2m_rd_client.c [hunk 11]
```diff
@@ -328,8 +329,27 @@ static void do_bootstrap_reg_timeout_cb(struct lwm2m_message *msg)
...
 }
+
 #endif
 
+int engine_trigger_bootstrap(void)
+{
+#if defined(CONFIG_LWM2M_RD_CLIENT_SUPPORT_BOOTSTRAP)
+	if (!sm_is_registered()) {
+		/* Bootstrap is not possible to trig */
+		LOG_WRN("Not Possible to Trig Bootstrap from state %u ", client.engine_state);
```
Comment (SeppoTakalo) [inline subsys/net/lib/lwm2m/lwm2m_rd_client.c:340]: `Trig` is not a English word. Also comment is irrelevant if log message is telling the same, so drop the comment.
I'll propose that you change this to
`Cannot trigger bootstrap from state %u`

File: subsys/net/lib/lwm2m/lwm2m_rd_client.c [hunk 12]
```diff
@@ -328,8 +329,27 @@ static void do_bootstrap_reg_timeout_cb(struct lwm2m_message *msg)
...
+		return -EPERM;
+	}
+
+	LOG_INF("Bootstrap initiated by server /1x/9 Exececute");
```
Comment (SeppoTakalo) [inline subsys/net/lib/lwm2m/lwm2m_rd_client.c:344]: I think the resource path can be left out, and there is a typo anyway.
So preferably change this to `Server Initiated Bootstrap`

File: subsys/net/lib/lwm2m/lwm2m_rd_client.c [hunk 13]
```diff
@@ -114,6 +114,7 @@ struct lwm2m_rd_client_info {
 	lwm2m_ctx_event_cb_t event_cb;
 
 	bool use_bootstrap : 1;
+
```
Comment (SeppoTakalo) [inline subsys/net/lib/lwm2m/lwm2m_rd_client.c:117]: unnecessary whitespace change.

**Review by SeppoTakalo — APPROVED** [review]
Looks OK, but I left couple of comments regarding the language on log messages.

File: subsys/net/lib/lwm2m/lwm2m_rd_client.c (hunk 11 above)
Comment (jheiskan81) [reply subsys/net/lib/lwm2m/lwm2m_rd_client.c:340]: Ok. I will fix that.

File: subsys/net/lib/lwm2m/lwm2m_rd_client.c (hunk 12 above)
Comment (jheiskan81) [reply subsys/net/lib/lwm2m/lwm2m_rd_client.c:344]: OK I will change.

File: subsys/net/lib/lwm2m/lwm2m_rd_client.c (hunk 13 above)
Comment (jheiskan81) [reply subsys/net/lib/lwm2m/lwm2m_rd_client.c:117]: Uuh There is a second one also. Good spot

File: subsys/net/lib/lwm2m/lwm2m_obj_server.c [hunk 14]
```diff
@@ -21,20 +21,42 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
...
+#define SERVER_TRANSPORT_BINDING_ID			7
+#define SERVER_REG_UPDATE_TRIGGER_ID			8
+#if defined(CONFIG_LWM2M_SERVER_OBJECT_VERSION_1_1)
+#define SERVER_BOOTSTRAP_UPDATE_TRIGGER_ID		9
+#define SERVER_APN_LINK_ID				10
+#define SERVER_TLS_DTLS_ALERT_CODE_ID			11
+#define SERVER_LAST_BOOTSTRAPPED_ID			12
+#define SERVER_REGISTRATION_PRIORITY_ORDER_ID		13
+#define SERVER_INITIAL_REGISTRATION_DELAY_TIMER_ID	14
+#define SERVER_REGISRATION_FAILURE_BLOCK_ID		15
```
Comment (jarlamsa) [inline subsys/net/lib/lwm2m/lwm2m_obj_server.c:49]: Nitpick: typo should be `SERVER_REGISTRATION_FAILURE_BLOCK_ID`

File: subsys/net/lib/lwm2m/lwm2m_obj_server.c [hunk 15]
```diff
@@ -78,6 +100,24 @@ static struct lwm2m_engine_obj_field fields[] = {
...
 	OBJ_FIELD_DATA(SERVER_TRANSPORT_BINDING_ID, RW, STRING),
 	OBJ_FIELD_EXECUTE(SERVER_REG_UPDATE_TRIGGER_ID),
+#if defined(CONFIG_LWM2M_SERVER_OBJECT_VERSION_1_1)
+	OBJ_FIELD_EXECUTE(SERVER_BOOTSTRAP_UPDATE_TRIGGER_ID),
+	OBJ_FIELD_DATA(SERVER_APN_LINK_ID, RW_OPT, OBJLNK),
+	OBJ_FIELD_DATA(SERVER_TLS_DTLS_ALERT_CODE_ID, R_OPT, U8),
+	OBJ_FIELD_DATA(SERVER_LAST_BOOTSTRAPPED_ID, R_OPT, TIME),
+	OBJ_FIELD_DATA(SERVER_REGISTRATION_PRIORITY_ORDER_ID, W_OPT, U16),
+	OBJ_FIELD_DATA(SERVER_INITIAL_REGISTRATION_DELAY_TIMER_ID, W_OPT, U16),
+	OBJ_FIELD_DATA(SERVER_REGISRATION_FAILURE_BLOCK_ID, W_OPT, BOOL),
```
Comment (jarlamsa) [inline subsys/net/lib/lwm2m/lwm2m_obj_server.c:110]: Same typo here `SERVER_REGISTRATION_FAILURE_BLOCK_ID`

File: subsys/net/lib/lwm2m/lwm2m_obj_server.c [hunk 16]
```diff
@@ -240,6 +293,32 @@ static struct lwm2m_engine_obj_inst *server_create(uint16_t obj_inst_id)
...
+	INIT_OBJ_RES_EXECUTE(SERVER_BOOTSTRAP_UPDATE_TRIGGER_ID, res[index], i,
+			     bootstrap_trigger_cb);
+	INIT_OBJ_RES_OPTDATA(SERVER_APN_LINK_ID, res[index], i, res_inst[index], j);
+	INIT_OBJ_RES_OPTDATA(SERVER_TLS_DTLS_ALERT_CODE_ID, res[index], i, res_inst[index], j);
+	INIT_OBJ_RES_OPTDATA(SERVER_LAST_BOOTSTRAPPED_ID, res[index], i, res_inst[index], j);
+	INIT_OBJ_RES_OPTDATA(SERVER_REGISTRATION_PRIORITY_ORDER_ID, res[index], i, res_inst[index],
+			     j);
+	INIT_OBJ_RES_OPTDATA(SERVER_INITIAL_REGISTRATION_DELAY_TIMER_ID, res[index], i,
+			     res_inst[index], j);
+	INIT_OBJ_RES_OPTDATA(SERVER_REGISRATION_FAILURE_BLOCK_ID, res[index], i, res_inst[index],
```
Comment (jarlamsa) [inline subsys/net/lib/lwm2m/lwm2m_obj_server.c:306]: And here `SERVER_REGISTRATION_FAILURE_BLOCK_ID`

**Review by jarlamsa — APPROVED** [review]
LGTM, nitpick about one typo.

File: subsys/net/lib/lwm2m/lwm2m_obj_server.c (hunk 14 above)
Comment (jheiskan81) [reply subsys/net/lib/lwm2m/lwm2m_obj_server.c:49]: Ok, Thanks Jarno. I will fix this.

File: subsys/net/lib/lwm2m/lwm2m_obj_server.c (hunk 15 above)
Comment (jheiskan81) [reply subsys/net/lib/lwm2m/lwm2m_obj_server.c:110]: Ok, Thanks Jarno. I will fix this.

File: subsys/net/lib/lwm2m/lwm2m_obj_server.c (hunk 16 above)
Comment (jheiskan81) [reply subsys/net/lib/lwm2m/lwm2m_obj_server.c:306]: Ok. I gonna fix that also.

**Comment by jheiskan81** [comment]
@SeppoTakalo , @VeijoPesonen  and @jarlamsa  Ihave been fix your comment and  PR shuold be ready for Merge.

File: subsys/net/lib/lwm2m/Kconfig [hunk 17]
```diff
@@ -23,6 +23,20 @@ config LWM2M_DTLS_SUPPORT
...
+	prompt "LwM2M Server object version"
+	default LWM2M_SERVER_OBJECT_VERSION_1_0
+	help
+	  Select which version of the Server object should be used.
+
+config LWM2M_SERVER_OBJECT_VERSION_1_0
+	bool "Server object version 1.0"
+
+config LWM2M_SERVER_OBJECT_VERSION_1_1
+	bool "Server object version 1.1 [EXPERIMENTAL]"
```
Comment (VeijoPesonen) [inline subsys/net/lib/lwm2m/Kconfig:36]: You should add `select EXPERIMENTAL` so that it will generate a warning during build time.

Comment (jheiskan81) [reply subsys/net/lib/lwm2m/Kconfig:36]: Ah OK I need to check

**Review by VeijoPesonen — APPROVED** [review]

**Review by SeppoTakalo — APPROVED** [review]

**Review by jarlamsa — APPROVED** [review]
